rust/src/write/writer.rs

```diff
@@ -0,0 +1,580 @@
+            Err(e) => {
+                let new_cursor = cursor_from_bytes(current_cursor_bytes.as_slice())?;
+                let _ = std::mem::replace(&mut self.cursor, new_cursor.clone());
+                let arrow_writer = new_underlying_writer(
+                    new_cursor,
+                    self.arrow_schema.clone(),
+                    self.writer_properties.clone(),
+                )?;
+                let _ = std::mem::replace(&mut self.arrow_writer, arrow_writer);
+                // TODO we used to clear partition values here, but since we pre-partition the
+                // record batches, we should never try with mismatching partition values.
+
+                Err(e.into())
+            }
+        }
+    }
+
+    /// Returns the current byte length of the in memory buffer.
+    /// This may be used by the caller to decide when to finalize the file write.
+    pub fn buffer_len(&self) -> usize {
+        self.cursor.len()
+    }
+}
+
+/// Partition a RecordBatch along partition columns
+pub fn divide_by_partition_values(
+    arrow_schema: ArrowSchemaRef,
+    partition_columns: Vec<String>,
+    values: &RecordBatch,
+) -> Result<Vec<PartitionResult>, DeltaWriterError> {
+    let mut partitions = Vec::new();
+
+    if partition_columns.is_empty() {
+        partitions.push(PartitionResult {
+            partition_values: HashMap::new(),
+            record_batch: values.clone(),
+        });
+        return Ok(partitions);
+    }
+
+    let schema = values.schema();
+
+    // collect all columns in order relevant for partitioning
+    let sort_columns = partition_columns
+        .clone()
+        .into_iter()
+        .map(|col| SortColumn {
+            values: values.column(schema.index_of(&col).unwrap()).clone(),
+            options: None,
+        })
+        .collect::<Vec<_>>();
+
+    let indices = lexsort_to_indices(sort_columns.as_slice(), None).unwrap();
+    let sorted_partition_columns = sort_columns
+        .iter()
+        .map(|c| SortColumn {
+            values: take(c.values.as_ref(), &indices, None).unwrap(),
+            options: None,
+        })
+        .collect::<Vec<_>>();
+
+    let partition_ranges = lexicographical_partition_ranges(sorted_partition_columns.as_slice())?;
+
+    for range in partition_ranges {
+        // get row indices for current partition
+        let idx: UInt32Array = (range.start..range.end)
+            .map(|i| Some(indices.value(i)))
+            .into_iter()
+            .collect();
+
+        let partition_key_iter = sorted_partition_columns.iter().map(|c| {
+            stringified_partition_value(&c.values.slice(range.start, range.end - range.start))
+                .unwrap()
+        });
+
+        let mut partition_values = HashMap::new();
+        for (key, value) in partition_columns.clone().iter().zip(partition_key_iter) {
+            partition_values.insert(key.clone(), value);
+        }
+
+        let batch_data = arrow_schema
```

Comment:
I'm not entirely sure how to interpret this line form the Delta protocol. 

> Values for all partition columns present in the schema MUST be present for all files in the table.

Does this mean, the partition values should not be omitted in the data files? We are writing them in the log, so they are somehow available, but right now not written in the data files. The previous implementation did write the data.

Reply:
Interesting, I was under the impression that we won't need need to duplicate partition column values in the data files. Let me double check this with the reference implementation tomorrow.

Reply:
I double checked and can confirm that partition values should not be written out into the data file. This sentence in the spec is a little bit mislead to me at least. Worth filing an upstream ticket to further clarify it.